### PR TITLE
Igal/mgmt 1481

### PR DESCRIPTION
**Pull request by tsorya**
1. Adding new api that sets cluster installed state
2. Adding new state "finalizing" (previous installed)

Finalizing state is set when 3 masters got installed

### CODE REVIEW

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -1884,6 +1884,47 @@ func (b *bareMetalInventory) ResetCluster(ctx context.Context, params installer.
...
+		}
+		if r := recover(); r != nil {
+			log.Error("failed to set complete cluster installation")
+			tx.Rollback()
+		}
+	}()
+
+	if tx.Error != nil {
+		log.WithError(tx.Error).Errorf("failed to start db transaction")
+		return installer.NewCompleteInstallationInternalServerError().WithPayload(
```
Comment (filanov) [inline internal/bminventory/inventory.go:1909]: you can use `common.NewApiError()`  relevant for all error replays

File: internal/cluster/cluster.go [hunk 2]
```diff
@@ -266,6 +272,25 @@ func (m *Manager) ResetCluster(ctx context.Context, c *common.Cluster, reason st
 	return nil
 }
 
+func (m *Manager) CompleteInstallation(ctx context.Context, c *common.Cluster, successfullyFinished bool, failureReason string, db *gorm.DB) error {
```
Comment (filanov) [inline internal/cluster/cluster.go:275]: it changes state, so i suggest to make it a transition and add it to the state machine.

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -1884,6 +1884,47 @@ func (b *bareMetalInventory) ResetCluster(ctx context.Context, params installer.
...
 
+func (b *bareMetalInventory) CompleteInstallation(ctx context.Context, params installer.CompleteInstallationParams) middleware.Responder {
+	log := logutil.FromContext(ctx, b.log)
+
+	log.Infof("complete cluster %s installation", params.ClusterID)
+
+	var c common.Cluster
+
+	txSuccess := false
+	tx := b.db.Begin()
```
Comment (filanov) [inline internal/bminventory/inventory.go:1895]: Not sure why a transaction is needed?

File: swagger.yaml [hunk 4]
```diff
@@ -446,6 +446,41 @@ paths:
           schema:
             $ref: '#/definitions/error'
 
+  /clusters/{cluster_id}/actions/complete_installation:
```
Comment (filanov) [inline swagger.yaml:449]: why isn't it part of update progress? what is the difference?

File: internal/cluster/finalizing.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+var _ StateAPI = (*Manager)(nil)
+
+func (i *finalizingState) RefreshStatus(ctx context.Context, c *common.Cluster, db *gorm.DB) (*UpdateReply, error) {
+	return &UpdateReply{
+		State:     clusterStatusFinalizing,
+		IsChanged: false,
+	}, nil
+}
+
+//func (i *finalizingState) Install(ctx context.Context, c *common.Cluster) (*UpdateReply, error) {
```
Comment (tsorya) [inline internal/cluster/finalizing.go:30]: delete

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (tsorya) [reply internal/bminventory/inventory.go:1895]: changing field in cluster, wanted to do it in transaction.

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (tsorya) [reply internal/bminventory/inventory.go:1909]: as i see common.NewApiError() is returned from help functions not from the api one

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (filanov) [reply internal/bminventory/inventory.go:1895]: Still don't understand why, you are doing this operation with `state = ?` and you are doing one operation. So i think you don't need it.

Comment (tsorya) [reply internal/bminventory/inventory.go:1895]: as we talked , update state needs transaction

File: swagger.yaml (hunk 4 above)
Comment (tsorya) [reply swagger.yaml:449]: we don't have update progress for cluster

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (filanov) [reply internal/bminventory/inventory.go:1909]: then you can use `GenerateErrorResponderWithDefault`

File: swagger.yaml (hunk 4 above)
Comment (filanov) [reply swagger.yaml:449]: @avishayt why do you think is the right solution?

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (filanov) [reply internal/bminventory/inventory.go:1895]: still don't understand why.

File: internal/cluster/cluster_test.go [hunk 6]
```diff
@@ -141,13 +141,13 @@ var _ = Describe("cluster monitor", func() {
...
 		})
-		It("installing -> installed", func() {
+		It("installing -> finalizing", func() {
 			createHost(id, "installed", db)
 			createHost(id, "installed", db)
 			createHost(id, "installed", db)
 
 			shouldHaveUpdated = true
-			expectedState = "installed"
+			expectedState = "finalizing"
```
Comment (filanov) [inline internal/cluster/cluster_test.go:150]: `models.ClusterStatus...`

File: internal/cluster/common.go [hunk 7]
```diff
@@ -17,6 +17,7 @@ const (
 	clusterStatusReady                  = "ready"
 	clusterStatusPrepareForInstallation = "preparing-for-installation"
 	clusterStatusInstalling             = "installing"
+	clusterStatusFinalizing             = "finalizing"
```
Comment (filanov) [inline internal/cluster/common.go:20]: i prefer to use the auto generated `models.ClusterStatus...` we should stop using those consts.

File: internal/cluster/common.go [hunk 8]
```diff
@@ -68,7 +70,13 @@ func updateState(state string, statusInfo string, c *common.Cluster, db *gorm.DB
 func updateClusterStateWithParams(log logrus.FieldLogger, srcStatus, statusInfo string, c *common.Cluster, db *gorm.DB,
 	extra ...interface{}) error {
 
-	updates := map[string]interface{}{"status": swag.StringValue(c.Status), "status_info": statusInfo}
+	updates := map[string]interface{}{"status": swag.StringValue(c.Status), "status_info": statusInfo, "status_updated_at": strfmt.DateTime(time.Now())}
```
Comment (filanov) [inline internal/cluster/common.go:73]: i don't thing that this code belongs here. 
it should be handled in the transition.

File: internal/cluster/transition_test.go [hunk 9]
```diff
@@ -61,6 +61,48 @@ var _ = Describe("CancelInstallation", func() {
 		Expect(swag.StringValue(c.StatusInfo)).Should(Equal("original error"))
 	})
 
+	It("complete installation success", func() {
```
Comment (filanov) [inline internal/cluster/transition_test.go:64]: those tests are not related to cancel installation. https://github.com/filanov/bm-inventory/pull/400/files#diff-d40fb42e8651cb3ce91fcd4b42d7697eR17

File: subsystem/cluster_test.go [hunk 10]
```diff
@@ -431,9 +435,44 @@ var _ = Describe("cluster install", func() {
...
+			waitForClusterState(ctx, clusterID, "finalizing", defaultWaitForClusterStateTimeout, "Finalizing cluster installation")
+			success := true
+			_, err = bmclient.Installer.CompleteInstallation(ctx,
+				&installer.CompleteInstallationParams{ClusterID: clusterID, CompletionParams: &models.CompletionParams{IsSuccess: &success, ErrorInfo: ""}})
+			Expect(err).NotTo(HaveOccurred())
+
 			waitForClusterState(ctx, clusterID, "installed", defaultWaitForClusterStateTimeout, "installed")
 		})
 
+		It("[only_k8s]install_cluster fail", func() {
```
Comment (filanov) [inline subsystem/cluster_test.go:447]: please add steps to this test, it will make it clearer `By(...)`

File: docs/ClusterState.dot [hunk 11]
```diff
@@ -14,13 +14,16 @@ digraph ClusterStateMachine {
...
+    finalizing -> installed [label = "installation\ncompleted", color=chartreuse4, fontcolor=chartreuse4];
+    finalizing -> error [label = "installation\nerror\nor canceled", color=crimson, fontcolor=crimson];
     error -> insufficient [label = "reset\nstate", color=lightpink3, fontcolor=lightpink3];
 
     {rank=min; start}
     {rank=max; installed}
     {rank=same; start; insufficient}
     {rank=same; ready; installing; error}
+    {rank=same; installing; finalizing; error}
     {rank=same; installed}
```
Comment (avishayt) [inline docs/ClusterState.dot:28]:
```suggestion
    {rank=same; installing; error}
    {rank=same; installed; finalizing}
```

File: docs/ClusterState.dot [hunk 12]
```diff
@@ -14,13 +14,16 @@ digraph ClusterStateMachine {
     ready -> insufficient [label = "min reqs\nnot met", color=lightpink3, fontcolor=lightpink3];
     error -> insufficient [label = "reset", color=lightpink3, fontcolor=lightpink3];
     ready -> installing [label = "installation started", color=darkolivegreen4, fontcolor=darkolivegreen4];
-    installing -> installed [label = "installation\ncompleted", color=chartreuse4, fontcolor=chartreuse4];
+    installing -> finalizing [label = "installation\n3 masters installed", color=chartreuse4, fontcolor=chartreuse4];
```
Comment (avishayt) [inline docs/ClusterState.dot:17]:
```suggestion
    installing -> finalizing [label = "3 masters\ninstalled", color=chartreuse4, fontcolor=chartreuse4];
```

File: docs/ClusterState.dot (hunk 11 above)
Comment (avishayt) [reply docs/ClusterState.dot:28]: (Regenerate the png file too)

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (tsorya) [reply internal/bminventory/inventory.go:1909]: just point me to example

File: internal/cluster/common.go (hunk 7 above)
Comment (tsorya) [reply internal/cluster/common.go:20]: i think it better to be consistence, if we have all those consts the new one must be added too. We need to move them altogether

File: internal/cluster/common.go (hunk 8 above)
Comment (tsorya) [reply internal/cluster/common.go:73]: this is the code that updates status, so not sure why to create one more function that will have 90% same code

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (filanov) [reply internal/bminventory/inventory.go:1909]: ctrl + f?

File: internal/cluster/common.go (hunk 7 above)
Comment (filanov) [reply internal/cluster/common.go:20]: then we will never stop using them and we should stop it. so please start from here.

File: internal/cluster/common.go (hunk 8 above)
Comment (filanov) [reply internal/cluster/common.go:73]: sorry i was talking about the next line:
```
if srcStatus == clusterStatusReady && swag.StringValue(c.Status) == clusterStatusPrepareForInstallation {
		updates["install_started_at"] = strfmt.DateTime(time.Now())
	} else if srcStatus == clusterStatusFinalizing && swag.StringValue(c.Status) == clusterStatusInstalled {
		updates["install_completed_at"] = strfmt.DateTime(time.Now())
	}
```

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (tsorya) [reply internal/bminventory/inventory.go:1895]: i think when u change field that can be changed from another process it is better to use transaction

File: internal/cluster/common.go (hunk 7 above)
Comment (tsorya) [reply internal/cluster/common.go:20]: don't agree but as you wish

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (filanov) [reply internal/bminventory/inventory.go:1902]: if you still insist on transaction you can use http://gorm.io/docs/transactions.html#Transactions

File: internal/cluster/common.go [hunk 13]
```diff
@@ -25,6 +25,7 @@ const (
 	statusInfoReady                           = "Cluster ready to be installed"
 	statusInfoInsufficient                    = "cluster is insufficient, exactly 3 known master hosts are needed for installation"
 	statusInfoInstalling                      = "Installation in progress"
+	statusFinalizing                          = "Finalizing cluster installation"
```
Comment (filanov) [inline internal/cluster/common.go:28]:
```suggestion
	statusInfoFinalizing                          = "Finalizing cluster installation"
```

File: internal/cluster/common.go [hunk 14]
```diff
@@ -45,7 +46,7 @@ func updateState(state string, statusInfo string, c *common.Cluster, db *gorm.DB
 	updates := map[string]interface{}{"status": state, "status_info": statusInfo, "status_updated_at": strfmt.DateTime(time.Now())}
 	if *c.Status == clusterStatusReady && state == clusterStatusPrepareForInstallation {
 		updates["install_started_at"] = strfmt.DateTime(time.Now())
-	} else if *c.Status == clusterStatusInstalling && state == clusterStatusInstalled {
+	} else if *c.Status == models.ClusterStatusFinalizing && state == clusterStatusInstalled {
```
Comment (filanov) [inline internal/cluster/common.go:49]: this part is not needed because you handle it in the state machine no?

File: internal/cluster/cluster.go [hunk 15]
```diff
@@ -217,10 +223,10 @@ func (m *Manager) GetCredentials(c *common.Cluster) (err error) {
 
 func (m *Manager) UploadIngressCert(c *common.Cluster) (err error) {
 	clusterStatus := swag.StringValue(c.Status)
-	if clusterStatus != clusterStatusInstalled {
-		err = errors.Errorf("Cluster %s is in %s state, upload ingress ca can be done only in installed state", c.ID, clusterStatus)
+	allowedStatuses := []string{models.ClusterStatusFinalizing, clusterStatusInstalled}
+	if !funk.ContainsString(allowedStatuses, clusterStatus) {
+		err = errors.Errorf("Cluster %s is in %s state, upload ingress ca can be done only in %s or %s state", c.ID, clusterStatus, models.ClusterStatusFinalizing, clusterStatusInstalled)
```
Comment (filanov) [inline internal/cluster/cluster.go:228]: you can return `APIError`
